experimental/lib/Support/BooleanLogic/BoolExpression.cpp

```diff
@@ -0,0 +1,358 @@
+  std::set<std::string> s;
+  getVariablesRec(s);
+  return s;
+}
+
+//---------Generating Truth Table based on Mintems (SOP Only)-----------
+
+void dynamatic::experimental::boolean::replaceDontCaresRec(
+    std::string s, std::set<std::string> &minterms) {
+  size_t dontCareIndex = s.find('d');
+  if (dontCareIndex == std::string::npos) {
+    minterms.insert(s);
+    return;
+  }
+  s[dontCareIndex] = '0';
+  replaceDontCaresRec(s, minterms);
+  s[dontCareIndex] = '1';
+  replaceDontCaresRec(s, minterms);
+}
+
+std::set<std::string> dynamatic::experimental::boolean::replaceDontCares(
+    const std::set<std::string> &minterms) {
+  std::set<std::string> mintermsWithoudDontCares;
+  for (StringRef s : minterms)
+    replaceDontCaresRec(s.str(), mintermsWithoudDontCares);
+
+  return mintermsWithoudDontCares;
+}
+
+// 'd': don't care, 'n':null/void
+// If a variable is a don't care: it should be set to 0 if it's negated and to 1
+// if it's non-negeated (for the minterm to be 1) If a variable is set to 0: if
+// it's present in non-negated form in the minterm, then the minterm requires
+// the variable to be both 0 and 1 for it to evaluate to 1, hence the minterm is
+// void (no setting of the values gives a 1) Similarly if a variable is set to 1
+// and is present in negated form in the minterm
+void BoolExpression::generateMintermVariable(
+    std::string &s, std::map<StringRef, int> varIndex) {
+  SingleCond *singleCond = static_cast<SingleCond *>(this);
+  char &c = s[varIndex[singleCond->id]];
+  if (c == 'd')
+    c = singleCond->isNegated ? '0' : '1';
+  else if (c == '0')
+    c = singleCond->isNegated ? '0' : 'n';
+  else if (c == '1')
+    c = (!singleCond->isNegated) ? '1' : 'n';
+}
+
+void BoolExpression::generateMintermAnd(
+    std::string &s, const std::map<StringRef, int> &varIndex) {
+  if (type == ExpressionType::VARIABLE) {
+    generateMintermVariable(s, varIndex);
+  } else if (type == ExpressionType::ZERO) { // 0 . exp =0 -> not a a minterm
+    s[0] = 'n';
+  } else if (type == ExpressionType::AND || type == ExpressionType::OR) {
+    Operator *op = static_cast<Operator *>(this);
+    if (op->left)
+      op->left->generateMintermAnd(s, varIndex);
+    if (op->right)
+      op->right->generateMintermAnd(s, varIndex);
+  }
+}
+
+void BoolExpression::generateMintermsOr(
+    unsigned numOfVariables, const std::map<StringRef, int> &varIndex,
+    std::set<std::string> &minterms) {
+  if (type == ExpressionType::OR) {
+    Operator *op = static_cast<Operator *>(this);
+    if (op->left)
+      op->left->generateMintermsOr(numOfVariables, varIndex, minterms);
+    if (op->right)
+      op->right->generateMintermsOr(numOfVariables, varIndex, minterms);
+  } else if (type == ExpressionType::ONE) { // 1 + exp = 1;
+    std::string s(numOfVariables, 'd');
+    minterms.insert(s);
+  } else if (type == ExpressionType::AND) {
+    std::string s(numOfVariables, 'd'); // initializing s
```

Comment:
This kind of comment is really not useful, it's immediately obvious from the line that s is being constructed. In-code comments are great but they should add meaning to otherwise hard to understand things, and generally focus more on *why* we do something rather than *what* the thing does (which, if the code is well written, should already be clear).